scheduler/src/cook/mesos/data_locality.clj

```diff
@@ -1,32 +1,118 @@
+     The number of jobs is limited by the configured batch size."
+    [db]
+    (let [{:keys [batch-size]} (config/data-local-fitness-config)
+          pending-jobs (->> db
+                            util/get-pending-job-ents
+                            (filter :job/data-local))
+          last-update-time-val @last-update-time
+          have-data?->job (group-by (fn [job] (contains? last-update-time-val (:job/uuid job)))
+                                    pending-jobs)
+          missing-data (->> (get have-data?->job false [])
+                            (sort-by :job/submit-time)
+                            (map :job/uuid))
+          sorted-have-data (->> (get have-data?->job true [])
+                                (map :job/uuid)
+                                (sort-by last-update-time-val))
+          pending-job-uuids (->> pending-jobs
+                                 (map :job/uuid)
+                                 (into (hash-set)))
+          no-longer-waiting (set/difference (into (hash-set) (keys last-update-time-val))
+                                            pending-job-uuids)]
+      {:to-fetch (into [] (take batch-size (concat missing-data sorted-have-data)))
+       :to-remove no-longer-waiting})))
+
+(defn fetch-data-local-costs
+  "Contacts the server to obtain the data local costs for the given job ids"
+  [job-ids]
+  (let [{:keys [cost-endpoint]} (config/data-local-fitness-config)
+        request {:batch (UUID/randomUUID)
+                 :tasks job-ids}
+        _ (log/debug "Updating data local costs for: " job-ids)
+        {:keys [body]} (http/post cost-endpoint {:body (cheshire/generate-string request)
+                                                 :content-type :json
+                                                 :accept :json
+                                                 :as :json-string-keys
+                                                 :spnego-auth true})
+        costs (body "costs")]
+    (->> costs
+         (map (fn [{:strs [task_id node_costs]}]
+                (let [costs (->> node_costs
+                                 (map (fn [{:strs [node cost]}]
+                                        [node cost]))
+                                 (into {}))]
+                  [task_id costs])))
+         (into {}))))
+
+(defn fetch-and-update-data-local-costs
+  [db]
+  (let [{:keys [to-fetch to-remove]} (job-ids-to-update db)]
+    (when (not (empty? to-fetch))
+      (log/debug "Updating data local costs for " to-fetch)
+      (let [new-costs (fetch-data-local-costs to-fetch)]
+        (log/debug "Got updated costs " new-costs)
+        (update-data-local-costs new-costs to-remove)))))
+
+(defn start-update-cycles!
+  [datomic-conn trigger-chan]
+  (log/info "Starting data local service update chan")
+  (util/chime-at-ch trigger-chan
+                    (fn [] (fetch-and-update-data-local-costs (d/db datomic-conn)))
```

Comment:
There are no `log/info` in the happy path that would tell us that the data locality daemon ran and updated something.